src/task.c

```diff
@@ -424,18 +424,19 @@ static void ctx_switch(jl_task_t *lastt)
         else
 #endif
         *pt = NULL; // can't fail after here: clear the gc-root for the target task now
-        lastt->ptls = NULL;
     }
 
     // set up global state for new task and clear global state for old task
     t->ptls = ptls;
     jl_atomic_store_relaxed(&ptls->current_task, t);
     JL_GC_PROMISE_ROOTED(t);
+    jl_signal_fence();
+    jl_set_pgcstack(&t->gcstack);
+    jl_signal_fence();
     lastt->ptls = NULL;
 #ifdef MIGRATE_TASKS
     ptls->previous_task = lastt;
 #endif
-    jl_set_pgcstack(&t->gcstack);
```

Comment:
In a debugging session today, Jameson figured out that the request was not delivered because we are hitting `if (ptls == NULL)` branch in `usr2_handler`

https://github.com/JuliaLang/julia/blob/569d56fa14cede9271a6ea93b030ca83fe08043a/src/signals-unix.c#L469-L471

It then would "look like" the signal is lost (but actually we were just ignoring it).

The fix was to ensure setting `lastt->ptls = NULL` *after* `jl_set_pgcstack(&t->gcstack)`. Since `lastt` is still the current task until `jl_set_pgcstack(&t->gcstack)` takes effect, we were previously observing `jl_get_current_task()->ptls == NULL`. This is fixed by the above patch.